File: backend-java/src/main/java/com/rpacloud/execution/process/ProcessOutputParser.java
package com.rpacloud.execution.process;

import java.util.Map;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.extern.slf4j.Slf4j;

/**
 * Extracts JSON payload from subprocess stdout/stderr.
 * 3-strategy best-effort parsing mirrors Python process_output_parser.py.
 */
@Slf4j
public final class ProcessOutputParser {

    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final TypeReference<Map<String, Object>> MAP_TYPE = new TypeReference<>() {};

    private ProcessOutputParser() {}

    public static Map<String, Object> extractJson(String output) {
        if (output == null || output.isBlank()) return null;
        String text = output.strip();

        // 1) Direct parse
        Map<String, Object> result = tryParse(text);
        if (result != null) return result;

        // 2) Last-line parse (reverse)
        String[] lines = text.split("\n");
        for (int i = lines.length - 1; i >= 0; i--) {
            String ln = lines[i].strip();
            if (ln.startsWith("{") && ln.endsWith("}")) {
                result = tryParse(ln);
                if (result != null) return result;
            }
        }

        // 3) Try from last '{' positions
        int idx = text.lastIndexOf('{');
        while (idx >= 0) {
            result = tryParse(text.substring(idx).strip());
            if (result != null) return result;
            idx = text.lastIndexOf('{', idx - 1);
        }

        return null;
    }

    private static Map<String, Object> tryParse(String json) {
        try {
            Object val = MAPPER.readValue(json, Object.class);
            if (val instanceof Map) {
                return MAPPER.readValue(json, MAP_TYPE);
            }
        } catch (Exception ignored) {}
        return null;
    }
}

File: backend/app/services/automation/process_output_parser.py
"""Utilities for parsing automation subprocess output.

Goal: improve result stability even when stdout/stderr contains noisy logs.
"""

from __future__ import annotations

import json
from typing import Any


def extract_json_payload(output: str) -> dict[str, Any] | None:
    """Extract a JSON object payload from process output.

    Strategy (best-effort):
    1) Try json.loads on whole output
    2) Try parsing line-by-line from bottom (common pattern: last line is JSON)
    3) Try parsing from last '{' occurrences (handles noisy prefixes)

    Returns:
        Parsed dict payload if found, otherwise None.
    """

    if not output:
        return None

    text = output.strip()
    if not text:
        return None

    # 1) Direct parse
    try:
        value = json.loads(text)
        if isinstance(value, dict):
            return value
    except Exception:
        pass

    # 2) Last-line parse (reverse)
    lines = [ln.strip() for ln in text.splitlines() if ln.strip()]
    for ln in reversed(lines):
        if not (ln.startswith("{") and ln.endswith("}")):
            continue
        try:
            value = json.loads(ln)
            if isinstance(value, dict):
                return value
        except Exception:
            continue

    # 3) Try from last '{' positions
    idx = text.rfind("{")
    while idx != -1:
        candidate = text[idx:].strip()
        try:
            value = json.loads(candidate)
            if isinstance(value, dict):
                return value
        except Exception:
            pass
        idx = text.rfind("{", 0, idx)

    return None
